lib/rules/prefer-const.js

```diff
@@ -37,6 +41,51 @@ function getWriteReferenceIfOnce(variable) {
 }
 
 /**
+ * Checks whether or not a given reference is in a loop condition or a
+ * for-loop's updater.
+ *
+ * @param {escope.Reference} reference - A reference to check.
+ * @returns {boolean} `true` if the reference is in a loop condition or a
+ *      for-loop's updater.
+ */
+function isInLoopHead(reference) {
+    var node = reference.identifier;
+    var parent = node.parent;
+    var assignment = false;
+
+    while (parent) {
+        if (LOOP_TYPES.test(parent.type)) {
+            return true;
+        }
+
+        // VariableDeclaration can be at ForInStatement.left
+        // This is catching the code like `for (const {b = ++a} of foo()) { ... }`
+        if (assignment &&
+            parent.type === "VariableDeclaration" &&
+            FOR_IN_OF_TYPES.test(parent.parent.type) &&
+            parent.parent.left === parent
+        ) {
+            return true;
```

Comment:
Wait, don't you mean the opposite? In the first case you should use const, and this rule should suggest it, but in the second case, you can't use const. Right?